database/database_test.go

```diff
@@ -298,3 +351,128 @@ func testMarkAsFailed(t *testing.T) {
 
 	// no need to mark them as succeeded, the other unit test covers that
 }
+
+// testCompatTransformSkylinkToHash is a unit test that verifies the correct
+// execution of the 'compatTransformSkylinkToHash' method on the database.
+func testCompatTransformSkylinkToHash(t *testing.T) {
+	// create context
+	ctx, cancel := context.WithTimeout(context.Background(), MongoDefaultTimeout)
+	defer cancel()
+
+	// create test database
+	db := newTestDB(ctx, t.Name())
+	defer db.Close()
+
+	// define a helper function to count the number of documents that the compat
+	// code should run on
+	countOldDocs := func() int {
+		// find all documents where the skyink has to be transformed to a hash
+		docs, err := db.find(ctx, bson.D{{"$and", []interface{}{
+			bson.D{{"skylink", bson.M{"$exists": true}}},
+			bson.D{{"$or", []interface{}{
+				bson.D{{"hash", nil}},
+				bson.D{{"hash", bson.M{"$exists": false}}},
+				bson.D{{"hash", emptyHash}},
+			}}},
+		}}})
+		if err != nil {
+			t.Fatal(err)
+		}
+		return len(docs)
+	}
+
+	// define a helper function to decode a skylink as string into a skylink obj
+	skylinkFromString := func(skylink string) (sl skymodules.Skylink) {
+		err := sl.LoadString(skylink)
+		if err != nil {
+			t.Fatal(err)
+		}
+		return
+	}
+
+	// assert the number of 'old' docs is 0
+	numOld := countOldDocs()
+	if numOld != 0 {
+		t.Fatalf("unexpected amount of docs, %v != 0", numOld)
+	}
+
+	// run the compat code on an empty database
+	err := db.compatTransformSkylinkToHash(ctx)
+	if err != nil {
+		t.Fatal("unexpected error", err)
+	}
+
+	// prepare 3 skylinks
+	sl1 := skylinkFromString("BAAWi3ou51qCH24Im0ESS-5_gKg60qGIYtta-ryrl1kBnQ")
+	sl2 := skylinkFromString("IABst6HgaJ0PIBMtmQ2qgH_wQlFg4bNnwAhff7DmJP6oyg")
+	sl3 := skylinkFromString("IABXaRBvjDTB3RizX3RfwdCoxt2Tff1buEXhlO7b9Unn8g")
+
+	// insert two documents with a skylink but no hash (like it was originally)
+	db.staticSkylinks.InsertOne(ctx, BlockedSkylink{
+		Skylink:        sl1.String(),
+		Reporter:       Reporter{},
+		Tags:           []string{"tag_1"},
+		TimestampAdded: time.Now().UTC(),
+	})
+	db.staticSkylinks.InsertOne(ctx, BlockedSkylink{
+		Skylink:        sl2.String(),
+		Reporter:       Reporter{},
+		Tags:           []string{"tag_1"},
+		TimestampAdded: time.Now().Add(time.Minute).UTC(),
+	})
+
+	// insert one documents with a skylink and a hash
+	db.staticSkylinks.InsertOne(ctx, BlockedSkylink{
+		Skylink:        sl3.String(),
+		Hash:           crypto.Hash(sl3.MerkleRoot()),
+		Reporter:       Reporter{},
+		Tags:           []string{"tag_1"},
+		TimestampAdded: time.Now().Add(time.Hour).UTC(),
+	})
+
+	// assert the number of 'old' docs is 2
```

Comment:
Don't we clear the `Skylink` field as part of the compat code?

Because `sl3` got both a hash and a skylink so in my opinion it should still be an `old` doc since we should clear the skylink from the db.

Reply:
As discussed on Discord, it makes more sense to deploy this in two stages. The PR documentation contains more information as to why I feel it makes for a safer and easier deploy/upgrade. TL;DR is that if we go in two stages, the database can perfectly well temporarily support both the old code and the new, which is always nice when you upgrade seeing as you can easily roll back if necessary.